`sync` -- explain algorithm in readme (#1600)
ava-labs/avalanchego: x/sync/README.md

When the client is created, it is given the root hash of the revision to sync to.
When it starts syncing, it requests from a server a range proof for the entire database.
The server replies with a range proof, which the client verifies.
If it's valid, the key-value pairs in the proof are written to the database.
If it's not, the client drops the proof and requests the proof from another server. 

A range proof sent by a server must return a continuous range of the key-value pairs, but may not 
return the full range that was requested.
For example, a client might request all the key-value pairs in [`requested_start`, `requested_end`] 
but only receive those in range [`requested_start`, `proof_end`] where `proof_end` < `requested_end`.
There might be too many key-value pairs to include in one message, or the server may be too busy to provide any more in its response.
Unless the database is very small, this means that the range proof the client receives in response to
 its range proof request for the entire database will not contain all of the key-value pairs in the database.

> If a client requests a range proof for range [`requested_start`, `requested_end`] but only receives 

Author: > Isn't it the case that range proofs during initial sync do not have a requested_end? And the first one doesn't have a requested_start either. This is alluded to when you say "[the client] requests from a server a range proof for the entire database".

Added a blurb that explains the omission of the lower/upper bounds.

> the client can determine if there are more keys because the computed partial hash does not match the Root Hash for this Revision.

Sorry I'm not sure what you mean by this -- would you mind elaborating? The root hash calculated from the trie resulting from the range proof has to match the root hash of the requested revision. Otherwise it's not valid.

> If a client requests a range proof for range [`requested_start`, `requested_end`] but only receives 

Reviewer: > > the client can determine if there are more keys because the computed partial hash does not match the Root Hash for this Revision.
> 
> Sorry I'm not sure what you mean by this -- would you mind elaborating? The root hash calculated from the trie resulting from the range proof has to match the root hash of the requested revision. Otherwise it's not valid.

What I mean is that if the client computes the root hash of what they have received so far and it does not match the `Root Hash` of the `Revision`, we know something is wrong -- we haven't received all the keys yet. Of course, this is an oversimplification of what actually happens, but logically it is still correct.

> a range proof for [`requested_start`, `proof_end`] where `proof_end` < `requested_end`

Reviewer: Our previous discussion might require some amendments here, since `proof_end` probably isn't provided so the comparison isn't possible in most range proof cases.

Author: Well, I suppose it depends on what your definition of `proof_end` is. If your definition of `proof_end` is "the greatest key such that the key's value or non-existence is shown by the proof" then all range proofs have `proof_end`. _However_, not all range proofs necessarily have and end proof (as defined by the `RangeProof` type).

it recognizes that it must still fetch all of the keys in [`proof_end`, `requested_end`]. 
It repeatedly requests range proofs for chunks of the remaining key range until it has all of the 
key-value pairs in [`requested_start`, `requested_end`].

> Note that the client may split the remaining key range into chunks and fetch chunks of key-value pairs in parallel.

Reviewer: ```suggestion
The client may split the remaining key range into chunks and fetch chunks of key-value pairs in parallel, possibly even from different servers.
```

Additional commits to the database may occur while the client is syncing.
The sync client can be notified that the root hash of the database it's trying to sync to has changed.
Detecting that the root hash to sync to has changed is done outside this package.
For example, if the database is being used to store blockchain state then the sync client would be 
notified when a new block is accepted because that implies a commit to the database.
If this occurs, the key-value pairs the client has learned about via range proofs may no longer be up to date.

> We use change proofs to correct the out of date key-value pairs. When the sync client is notified that

Reviewer: ```suggestion
We use change proofs as an optimization to correct the out of date key-value pairs. When the sync client is notified that
```

the root hash to sync to has changed, it requests a change proof from a server for a given key range.
For example, if a client has the key-value pairs in range [`start`, `end`] that were in the database 
when it had `root_hash`, then it will request a change proof that provides all of the key-value changes 
in range [`start`, `end`] from the database version with root hash `root_hash` to the database version with root hash `new_root_hash`.
The client verifies the change proof, and if it's valid, it applies the changes to its database.
If it's not, the client drops the proof and requests the proof from another server.

A server needs to have history in order to serve a change proof.
Namely, it needs to know all of the database changes between two roots.

> If the server does not have sufficient history to generate a change proof, it will send a range proof instead.

Reviewer: ```suggestion
If the server does not have sufficient history to generate a change proof, it will send a range proof of the `new_root_hash` instead.
```

### `RangeProof`

This message is sent from the server to the client in response to a `SyncGetRangeProofRequest`. It contains the key-value pairs that were in the requested key range when the database had the requested root, as well as a proof that the key-value pairs are correct.

### `SyncGetChangeProofRequest`

This message is sent from the client to the server to request a change proof between the given root IDs. That is, the client says, "Give me the key-value pairs that changed between the time the database had this root and that root."

### `SyncGetChangeProofResponse`

This message is sent from the server to the client in response to a `SyncGetChangeProofRequest`. If the server had sufficient history to generate a change proof, it contains a change proof that contains the key-value pairs that changed between the requested roots. If the server did not have sufficient history to generate a change proof, it contains a range proof that contains the key-value pairs that were in the database when the database had the latter root.

## Algorithm

> When the client starts with an empty database, it has no key-value pairs. It requests from a server a range proof for the entire database. The server replies with a range proof, which the client verifies. If it's valid, the key-value pairs in the proof are written to the database. If it's not, the client drops the proof and requests the proof from another server. 

Reviewer 2: > When the client starts with an empty database
The user may not start with a totally empty database, not sure if we want to discuss that case as well.

Author: removed reference to empty database since I don't think it's really relevant

A range proof sent by a server may be valid but not contain all of the key-value pairs in the requested range. For example, a client might request all the key-value pairs in [`start`, `end`] but only receive those in range [`start`, `end'`] where `end'` < `end`. There might be too many key-value pairs to include in one message, or the server may be too busy to provide any more in its response. Unless the database is very small, this means that the range proof the client receives in response to its range proof request for the entire database will not contain all of the key-value pairs in the database.

> For each key range, the sync client keeps track of the root ID of the last time it synced that key range. For example, it will store information that says something like, "I have all of the key-value pairs that were in range [`start`, `end`] when the database's root was `rootID`" for some keys `start` and `end`, and some database `rootID`.

Reviewer 2: I think this could use some clarity rewrites.  It technically only tracks it for each received proof rather than all key ranges.  It might also be worth specifying that it is the remote DB's Root ID that is recorded rather than whatever the local one is at the moment.

Author: >  It technically only tracks it for each received proof rather than all key ranges

Isn't this semantically the same, though?

> It might also be worth specifying that it is the remote DB's Root ID that is recorded rather than whatever the local one is at the moment. 

done

Reviewer 2: You could have data in your local db from before syncing started, so the known roots are only for received proofs during this sync cycle rather than from all time.

> If the first range proof contained all of the key-value pairs up to some key `end`, the client recognizes that it must fetch all of the keys after `end`. It divides the remaining key range into chunks, and repeats this process by asking servers for a range proof for each chunk until it has all of the key-value pairs in the database.

Reviewer 2: was this supposed to be end' rather than end?

Author: No but I guess it isn't entirely clear since you're asking lol

Reviewer 2: I guess the only clarity then would be to add that it only breaks things up if it detects there were keys between the end' and the end.  If the end' == end then it is done and no new work needs to be added.

Author: I rewrote this bit. Lmk what you think

> If the first range proof contained all of the key-value pairs up to some key `end`, the client recognizes that it must fetch all of the keys after `end`. It divides the remaining key range into chunks, and repeats this process by asking servers for a range proof for each chunk until it has all of the key-value pairs in the database.

Reviewer 2: > It divides the remaining key range into chunks,

not sure it is worth explicitly mentioning, but it doesn't always divide and it also prioritizes the first chunk over the second chunk so that ranges with matching rootID will grow rather than have many disjoint ranges.

Author: reworded it a bit. I think the exact algorithm for chunking/dechunking (those are words, right?) is too much detail for this document since a client could choose to implement it differently

However, the database may be changing as the client is syncing. The sync client can be notified that the root ID of the database it's trying to sync to has changed. Detecting that the root ID to sync to has changed is done outside this package. If this occurs, the key-value pairs the client has learned about via range proofs may no longer be valid.

We use change proofs to correct the out of date key-value pairs. When the sync client is notified that the root ID to sync to has changed, it requests a a change proof from a server for a given key range. For example, if a client has the key-value pairs in range [`start`, `end`] that were in the database when it had `rootID`, then it will request a change proof that provides all of the key-value changes in range [`start`, `end`] from the database version with root ID `rootID` to the database version with root ID `newRootID`. The client verifies the change proof, and if it's valid, it applies the changes to its database. If it's not, the client drops the proof and requests the proof from another server.

> A server needs to have history in order to serve a change proof. Namely, it needs to know all of the database changes between two roots. If the server does not have sufficient history to generate a change proof, it will send a range proof instead. The client will verify and apply the range proof. Note that change proofs, like range proofs, may not contain all of the key-value pairs in the requested range. This is OK because as mentioned above, the client tracks the root ID associated with each range of key-value pairs it has, so it knows which key-value pairs are out of date. If a change proof doesn't contain all of the key-value pairs in the requested range, the client will request another change proof for the remaining key-value pairs that are out of date.

Reviewer 2: might be worth clarifying that it should contain all the changed keys in [start, end'] and will then will request [end', end]

Author: done

# `sync` package

## Overview

> This package implements a client and server that allows for the syncing of a [MerkleDB](../merkledb/README.md). The servers have an up to date version of the database, and the clients have an out of date version of the database or an empty database.

Reviewer 3: Here in terms of syncing, we always refer to finalized blocks, right?

It's planned that these client and server implementations will eventually be compatible with Firewood.

## Messages

There are four message types sent between the client and server:

1. `SyncGetRangeProofRequest`
2. `RangeProof`
3. `SyncGetChangeProofRequest`

> 4. `SyncGetChangeProofResponse`

Reviewer 4: I hate how much `RangeProof` sticks out here

Reviewer 2: Would you hate it if we made a response struct that just had the range proof in it?

Author: We could rename it in the proto if you want, but if we do let's do it in a separate PR

These message types are defined in `avalanchego/proto/sync.proto`.
For more information on range proofs and change proofs, see their definitions in `avalanchego/merkledb/proof.go`.

### `SyncGetRangeProofRequest`

This message is sent from the client to the server to request a range proof for a given key range and root ID. That is, the client says, "Give me the key-value pairs that were in this key range when the database had this root."

### `RangeProof`

> This message is sent from the server to the client in response to a `SyncGetRangeProofRequest`. It contains the key-value pairs that were in the requested key range when the database had the requested root, as well as a proof that the key-value pairs are correct.

Reviewer 4: We should probably expand here that the server isn't required to return the full range. (I know we mention that later... but seems important in this part)

> This message is sent from the server to the client in response to a `SyncGetRangeProofRequest`. It contains the key-value pairs that were in the requested key range when the database had the requested root, as well as a proof that the key-value pairs are correct. If a server can't serve the entire requested key range in one response, its response will omit keys from the end of the range rather than the start. For example, if a client requests a range proof for range [`start`, `end`] but the server can't fit all the key-value pairs in one response, it'll send a range proof for [`start`, `end'`] where `end'` < `end`, as opposed to sending a range proof for [`start'`, `end`] where `start'` > `start`.

Author: Sorry I don't understand the question -- would you mind clarifying? The client never sends any acks to the server.

### `SyncGetChangeProofRequest`

This message is sent from the client to the server to request a change proof between the given root IDs. That is, the client says, "Give me the key-value pairs that changed between the time the database had this root and that root."

### `SyncGetChangeProofResponse`

> This message is sent from the server to the client in response to a `SyncGetChangeProofRequest`. If the server had sufficient history to generate a change proof, it contains a change proof that contains the key-value pairs that changed between the requested roots. If the server did not have sufficient history to generate a change proof, it contains a range proof that contains the key-value pairs that were in the database when the database had the latter root.

Reviewer 4: Same with the range proof - seems like the partial return should be documented here.

## Algorithm

> For each key range, the sync client keeps track of the root ID of the database revision for which it has downloaded that key range. For example, it will store information that says something like, "I have all of the key-value pairs that were in range [`start`, `end`] when the database's root was `rootID`" for some keys `start` and `end`, and some database `rootID`. Note that `rootID` is the root ID that the client is trying to sync to, not the root ID of its own incomplete database.

Reviewer: Let's use the terms as defined here: https://github.com/ava-labs/firewood/blob/main/README.md#termimology

`rootID` -> `Root Hash`
`database revision` -> `Revision`

The last sentence should be rephrased to something like:
> The hash of the incomplete database on a client is never sent anywhere, because it does not represent a Root Hash of any Revision.

Author: > rootID -> Root Hash

Done

> The hash of the incomplete database on a client is never sent anywhere, because it does not represent a Root Hash of any Revision.

Added

> When the client starts syncing, it requests from a server a range proof for the entire database. The server replies with a range proof, which the client verifies. If it's valid, the key-value pairs in the proof are written to the database. If it's not, the client drops the proof and requests the proof from another server. 

Author: Added "When the client is created, it is given the root hash of the revision to sync to" to clarify.

> A range proof sent by a server may be valid but not contain all of the key-value pairs in the requested range. For example, a client might request all the key-value pairs in [`start`, `end`] but only receive those in range [`start`, `end'`] where `end'` < `end`. There might be too many key-value pairs to include in one message, or the server may be too busy to provide any more in its response. Unless the database is very small, this means that the range proof the client receives in response to its range proof request for the entire database will not contain all of the key-value pairs in the database.

Reviewer 4: We need to write somewhere the word `continuous`.

> A range proof sent by a server may be valid but not contain all of the key-value pairs in the requested range.

sounds wrong.

```
A range proof sent by a server must return a continuous range of the key-value pairs, but may not return the full range that was requested.
```

Would be more clear imo

Author: I included the bit about validity since it's possible for the server to send the client an invalid range proof since it's a byzantine environment. Edited to mention that the client includes the maximum number of key-value pairs and maximum response size in its request. 

> Can you describe the current algorithm for how large of a Range Proof it currently returns? Is it complicated?

This is an implementation detail, so I'm not sure it makes sense to include it here, but basically the server tries to respond with the entire key-value range, but if the proof ends up being too large, it includes only a portion of the range.

> If a client requests a range proof for range [`start`, `end`] but only receives a range proof for [`start`, `end'`] where `end'` < `end` it recognizes that it must still fetch all of the keys in [`end`, `end'`]. It repeatedly requests range proofs for chunks of the remaining key range until it has all of the key-value pairs in [`start`, `end`]. Note that the client may split the remaining key range into chunks and fetch chunks of key-value pairs in parallel.

Reviewer 4: ```suggestion
If a client requests a range proof for range [`start`, `end`] but only receives a range proof for [`start`, `end'`] where `end'` < `end` it recognizes that it must still fetch all of the keys in [`end'`, `end`]. It repeatedly requests range proofs for chunks of the remaining key range until it has all of the key-value pairs in [`start`, `end`]. Note that the client may split the remaining key range into chunks and fetch chunks of key-value pairs in parallel.
```

Reviewer: nit: [`start`, `end`] -> [`requested_start`, `requested_end`], and the other [`start`, `end`] -> [`proof_start`, `proof_end`].

It was confusing to see `end` < `end` which seemed tautologically false, and makes reading this paragraph a lot harder.

I'd like to see more of an explanation of the "may" statements in here -- when does it actually split the remaining key ranges into chunks and when does it actually fetch chunks of key-value pairs in parallel?

Author: Edited to use `requested_end` and `proof_end` instead of `end` and `end'`, which I could see being hard to read.

> I'd like to see more of an explanation of the "may" statements in here -- when does it actually split the remaining key ranges into chunks and when does it actually fetch chunks of key-value pairs in parallel?

Again, this is an implementation detail, so I'm not sure it makes sense to explain it in more detail here. I could just omit mention of this altogether. What do you think?

> The database may be changing as the client is syncing. The sync client can be notified that the root ID of the database it's trying to sync to has changed. Detecting that the root ID to sync to has changed is done outside this package. If this occurs, the key-value pairs the client has learned about via range proofs may no longer be valid.

Reviewer 4: `valid` seems too harsh here (like an error has occurred)... super nit-picky.
```suggestion
The database may be changing as the client is syncing. The sync client can be notified that the root ID of the database it's trying to sync to has changed. Detecting that the root ID to sync to has changed is done outside this package. If this occurs, the key-value pairs the client has learned about via range proofs may no longer be up to date.
```

Reviewer: Change "[t]he database may be changing" (which implies a specific Revision is changing, which is impossible) to "additional `Commit`s may occur while the client is syncing"
Also `root ID` -> `Root Hash` (or `Revision` in some cases depending on what you're talking about)

If I want more information about how the new Revision detection happens, what package does that?

> We use change proofs to correct the out of date key-value pairs. When the sync client is notified that the root ID to sync to has changed, it requests a a change proof from a server for a given key range. For example, if a client has the key-value pairs in range [`start`, `end`] that were in the database when it had `rootID`, then it will request a change proof that provides all of the key-value changes in range [`start`, `end`] from the database version with root ID `rootID` to the database version with root ID `newRootID`. The client verifies the change proof, and if it's valid, it applies the changes to its database. If it's not, the client drops the proof and requests the proof from another server.

Reviewer 4: ```suggestion
We use change proofs to correct the out of date key-value pairs. When the sync client is notified that the root ID to sync to has changed, it requests a change proof from a server for a given key range. For example, if a client has the key-value pairs in range [`start`, `end`] that were in the database when it had `rootID`, then it will request a change proof that provides all of the key-value changes in range [`start`, `end`] from the database version with root ID `rootID` to the database version with root ID `newRootID`. The client verifies the change proof, and if it's valid, it applies the changes to its database. If it's not, the client drops the proof and requests the proof from another server.
```

We use change proofs to correct the out of date key-value pairs. When the sync client is notified that the root ID to sync to has changed, it requests a change proof from a server for a given key range. For example, if a client has the key-value pairs in range [`start`, `end`] that were in the database when it had `rootID`, then it will request a change proof that provides all of the key-value changes in range [`start`, `end`] from the database version with root ID `rootID` to the database version with root ID `newRootID`. The client verifies the change proof, and if it's valid, it applies the changes to its database. If it's not, the client drops the proof and requests the proof from another server.

> A server needs to have history in order to serve a change proof. Namely, it needs to know all of the database changes between two roots. If the server does not have sufficient history to generate a change proof, it will send a range proof instead. The client will verify and apply the range proof. Note that change proofs, like range proofs, may not contain all of the key-value pairs in the requested range. This is OK because as mentioned above, the client tracks the root ID associated with each range of key-value pairs it has, so it knows which key-value pairs are out of date. Similar to range proofs, if a client requests the changes in range [`start`, `end`], but the server replies with all of the changes in [`start`, `end'`] for some `end'` < `end`, the client will repeatedly request change proofs until it gets remaining key-value pairs (namely in [`end'`, `end`]). 

Reviewer: Does the client know that it sent a `RangeProof`? Seems like it has to know, because it should implicitly delete anything that isn't present in the `RangeProof` that it currently has.

I think you can avoid the repetition here and just mention that the `ChangeProof` might be incomplete, and the same algorithm is applied to fetch more changes as mentioned earlier. Actually it looks like you talk about that in the next paragraph.

Author: > Seems like it has to know, because it should implicitly delete anything that isn't present in the RangeProof that it currently has.

It doesn't necessarily need to delete everything in the old range proof, only the key-value pairs that are covered by the new range proof.

For example, if the client has a range proof for [`start_1`, `end_1`] at `root_hash_1`, and it gets a range proof for [`start_1`, `end_2`] at `root_hash_2` where `end_2` < `end_1` and `root_hash_2` is after `root_hash_1`, then it can still keep the key-values in [`end_2`, `end_1`] from the revision at `root_hash_1`. It'll just have to update those via change proof as well, since they're (potentially) stale.

> A server needs to have history in order to serve a change proof. Namely, it needs to know all of the database changes between two roots. If the server does not have sufficient history to generate a change proof, it will send a range proof instead. The client will verify and apply the range proof. Note that change proofs, like range proofs, may not contain all of the key-value pairs in the requested range. This is OK because as mentioned above, the client tracks the root ID associated with each range of key-value pairs it has, so it knows which key-value pairs are out of date. Similar to range proofs, if a client requests the changes in range [`start`, `end`], but the server replies with all of the changes in [`start`, `end'`] for some `end'` < `end`, the client will repeatedly request change proofs until it gets remaining key-value pairs (namely in [`end'`, `end`]). 

Reviewer 3: > If the server does not have sufficient history to generate a change proof, it will send a range proof instead. The client will verify and apply the range proof.

In such case, the client needs to delete all previous key-value pairs stored?

Author: No, the client doesn't need to do this. It will just update the key-value pairs in the range covered by the range proof.

## Overview

This package implements a client and server that allows for the syncing of a [MerkleDB](../merkledb/README.md).

> The servers have an up to date version of the database, and the clients have an out of date version of the database or an empty database.

Author: There is no gRPC server. Messages are sent through AvalancheGo's app messages

> The servers have an up to date version of the database, and the clients have an out of date version of the database or an empty database.

Reviewer 5: Aha OK makes more sense thx

## Algorithm

> For each proof it receives, the sync client tracks the root hash of the revision associated with the proof's key-value pairs. For example, it will store information that says something like, "I have all of the key-value pairs that are in range [`start`, `end`] for the revision with root `root_hash`" for some keys `start` and `end`. Note that `root_hash` is the root hash of the revision that the client is trying to sync to, not the root hash of its own (incomplete) database. Tracking the revision associated with each downloaded key range, as well as using data in its own (incomplete) database, allows the client to figure out which key ranges are not up to date and need to be synced. The hash of the incomplete database on a client is never sent anywhere because it does not represent a root hash of any revision.

Reviewer 5: would it be possible to do a hard wrap of like 80 or 100 chars? I find this nice for reviewing text in markdown.

Reviewer 4: I use this in my `vscode` settings for this kinda stuff - but don't treat it as a hard rule
```json
"editor.rulers": [
    80
],
```

Author: added newlines
